transformations/negate_strengthen/transformation.py

```diff
@@ -0,0 +1,462 @@
+import random
+from interfaces.SentenceOperation import SentenceAndTargetOperation
+from tasks.TaskTypes import TaskType
+
+import re
+import spacy
+from spacy.tokenizer import Tokenizer
+import pandas as pd
+import nltk
+from nltk.corpus import wordnet
+from nltk.stem.wordnet import WordNetLemmatizer
+from nltk.metrics import edit_distance
+from nltk import pos_tag
+from pattern.en import conjugate as conjugate_en
+from typing import Union
+
+# defaults
+_RE_COMBINE_WHITESPACE = re.compile(r"\s+")
+MODAL_TYPE_LEMMA = ['be', 'to', 'have', 'get', 'do']
+MODAL_STRENGTHEN_DICT = {
+    'could': 'would',
+    'should': 'would',
+    'would': 'would',
+    'can': 'will',
+    'may': 'will',
+    'might': 'will',
+    'will': 'will'
+}
+
+# intialise
+try:
+    conjugate_en(verb='testing',tense='present',number='singular')
+except:
+    pass
+
+
+class NegateStrengthen(SentenceAndTargetOperation):
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.SENTIMENT_ANALYSIS]
+    languages = ["en"]
+    tgt_languages = ["en"]
+
+    def __init__(self, seed=0, verbose=False):
+        super().__init__(seed)
+
+        # downloads
+        nltk.download('wordnet')
+        nltk.download('averaged_perceptron_tagger')
+
+        # initialise
+        self.verbose = verbose
+        self.nlp = spacy.load("en_core_web_sm")
+        self.nlp.tokenizer = Tokenizer(
+            self.nlp.vocab, 
+            prefix_search=re.compile('''^\\$[a-zA-Z0-9]''').search
+            )
+        self.wnl = WordNetLemmatizer()
+        self.seed = seed
+        random.seed(self.seed)
+
+
+    def text_to_spacy_table_info(self, _text):
+
+        doc = self.nlp(_text)
+
+        table = []
+        letters2word = {}
+        
+        # future to do: keep only relevant items
+        for idx, token in enumerate(doc):
+            letters2word[token.idx] = idx
+            table.append([token.text, token.lemma_, token.pos_, token.tag_, 
+                        token.dep_, token.head.text, token.head.idx, token.head.pos_])
+
+        table = pd.DataFrame(table)
+        table.columns = ['text', 'lemma', 'pos', 'pos_tag', 'dep', 'head_text', 'head_id', 'head_pos']
+        # convert idx from letter idx to word idx level
+        table['head_id'] = [letters2word[i] for i in table['head_id']]
+        if self.verbose:
+            print(table)
+
+        return table
+
+
+    def negation_rules(self, tgx, itemdict, text, pos, sentid2tid, method=[], \
+        num_tries=2, curr_try=0):
+
+        curr_try += 1
+        edit_id = tgx
+
+        if self.verbose:
+            print('before {}, actual {}, after {}'.format(pos[tgx-2], pos[tgx-1], pos[tgx]))
+        if curr_try > num_tries:
```

Comment:
Can we make 'num_tries' as a constant variable, instead of passing it every time (in your recursion & from text_to_negated_edits calls)?

Reply:
Ok, done